source/credentials_provider_sts.c

```diff
@@ -0,0 +1,668 @@
+}
+
+/* called upon aquiring a connection from the pool */
+static void s_on_connection_setup_fn(struct aws_http_connection *connection, int error_code, void *user_data) {
+    struct sts_creds_provider_user_data *provider_user_data = user_data;
+    struct aws_credentials_provider_sts_impl *provider_impl = provider_user_data->provider->impl;
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+        "(id=%p): connection returned with error code %d",
+        (void *)provider_user_data->provider,
+        error_code);
+
+    if (error_code) {
+        aws_raise_error(error_code);
+        goto error;
+    }
+    provider_user_data->connection = connection;
+
+    if (aws_byte_buf_init(&provider_user_data->output_buf, provider_impl->provider->allocator, 2048)) {
+        provider_impl->function_table->aws_http_connection_manager_release_connection(
+            provider_impl->connection_manager, connection);
+    }
+
+    struct aws_http_make_request_options options = {
+        .manual_window_management = false,
+        .user_data = user_data,
+        .request = provider_user_data->message,
+        .self_size = sizeof(struct aws_http_make_request_options),
+        .on_response_headers = NULL,
+        .on_response_header_block_done = NULL,
+        .on_response_body = s_on_incoming_body_fn,
+        .on_complete = s_on_stream_complete_fn,
+    };
+
+    struct aws_http_stream *stream =
+        provider_impl->function_table->aws_http_connection_make_request(connection, &options);
+    if (!stream) {
+        provider_impl->function_table->aws_http_connection_manager_release_connection(
+            provider_impl->connection_manager, connection);
+        goto error;
+    }
+
+    return;
+error:
+    provider_user_data->callback(NULL, provider_user_data->user_data);
+    s_clean_up_user_data(provider_user_data);
+}
+
+/* called once sigv4 signing is complete. This isn 't called if MTLS is used. */
+void s_on_signing_complete(struct aws_signing_result *result, int error_code, void *userdata) {
+    (void)result;
+    (void)error_code;
+    struct sts_creds_provider_user_data *provider_user_data = userdata;
+    struct aws_credentials_provider_sts_impl *sts_impl = provider_user_data->provider->impl;
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+        "(id=%p): signing completed with error code %d",
+        (void *)provider_user_data->provider,
+        error_code);
+
+    if (error_code) {
+        aws_raise_error(error_code);
+        goto error;
+    }
+
+    if (aws_apply_signing_result_to_http_request(
+            provider_user_data->message, provider_user_data->provider->allocator, result)) {
+        goto error;
+    }
+
+    sts_impl->function_table->aws_http_connection_manager_acquire_connection(
+        sts_impl->connection_manager, s_on_connection_setup_fn, provider_user_data);
+    return;
+
+error:
+    provider_user_data->callback(NULL, provider_user_data->user_data);
```

Comment:
It feels like the callback and the clean up could be combined into a single function that does both.  That's what the imds provider does I believe.